Added Tutorial: Paymaster Tutorial with API3 dAPIs (#575)
matter-labs/zksync-web-era-docs: docs/dev/tutorials/api3-usd-paymaster-tutorial.md

Within a paymaster, price oracles can be used to provide price data on-chain for execution.

For this paymaster tutorial, we will use dAPIs to get the price of [ETH/USD](https://market.api3.org/dapis/zksync-goerli-testnet/ETH-USD) and [USDC/USD](https://market.api3.org/dapis/zksync-goerli-testnet/USDC-USD) datafeeds and use it to calculate gas in USDC value so that users can pay for their transactions with USDC.

::: note
If you want to use an ERC20 token other than USDC, you can do so by changing the dAPIs used in the paymaster. For example, if you want to use DAI, you can use the [DAI/USD](https://market.api3.org/dapis/zksync-goerli-testnet/DAI-USD) dAPI instead of USDC/USD.
:::

## Project repo

The tutorial code is available [here](https://github.com/vanshwassan/zk-paymaster-dapi-poc)

## Set up the project

1. We're going to use [zkSync CLI](/docs/tools/zksync-cli/) to set up an empty project. Install it globally:

```sh

> $ yarn add global zksync-cli@latest

Author: Done 👍🏻

```

2. After installation, run the following command to create a new project:

```sh
$ yarn zksync-cli create paymaster-dapi
```

3. This will create a new zkSync project called `paymaster-dapi` with a basic `Greeter` contract. `cd` into the project directory:

```sh
$ cd paymaster-dapi
```

> 3. Add the project dependencies, including openzeppelin, zkSync and API3 contracts:

Author: Fixed

```

## Design

For the sake of simplicity, we will use a modified OpenZeppelin ERC20 implementation. For that, we are going to code a basic ERC20 token `mockUSDC` which will be used to pay for the transactions.

1. Create a new contract `mockUSDC.sol` under `/contracts` directory and add the following code:

```solidity
// SPDX-License-Identifier: UNLICENSED

pragma solidity ^0.8.8;

import "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract MyERC20 is ERC20 {
    uint8 private _decimals;

    constructor(
        string memory name_,
        string memory symbol_,
        uint8 decimals_
    ) ERC20(name_, symbol_) {
        _decimals = decimals_;
    }

    function mint(address _to, uint256 _amount) public returns (bool) {
        _mint(_to, _amount);
        return true;
    }

    function decimals() public view override returns (uint8) {
        return _decimals;
    }
}
```

Under contracts, you will find `Greeter.sol`. This is the contract that we will be using to test our paymaster to set a greeting message on-chain.

### Paymaster solidity contract

> 2. We can now create our paymaster contract `MyPaymaster.sol` under `/contracts` directory. It is a custom implementation of the zkSync paymaster contract that uses dAPIs.

Reviewer: Might be beneficial to start this section with a skeleton paymaster contract and then walk the audience through the required changes. It will be easier to follow and cause less confusion imo. Something like:

---

2. Open `MyPaymaster.sol` in your preferred text editor and paste the skeleton contract: 

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/token/ERC20/IERC20.sol";

import {IPaymaster, ExecutionResult, PAYMASTER_VALIDATION_SUCCESS_MAGIC} from "@matterlabs/zksync-contracts/l2/system-contracts/interfaces/IPaymaster.sol";
import {IPaymasterFlow} from "@matterlabs/zksync-contracts/l2/system-contracts/interfaces/IPaymasterFlow.sol";
import {TransactionHelper, Transaction} from "@matterlabs/zksync-contracts/l2/system-contracts/libraries/TransactionHelper.sol";

import "@matterlabs/zksync-contracts/l2/system-contracts/Constants.sol";

contract MyPaymaster is IPaymaster {
    
    modifier onlyBootloader() {
        require(msg.sender == BOOTLOADER_FORMAL_ADDRESS, "Only bootloader can call this method");
        // Continue execution if called from the bootloader.
        _;
    }

    constructor(address _erc20) {
        allowedToken = _erc20;
    }
    
    function setDapiProxy(
        address _USDCproxy,
        address _ETHproxy
    ) public onlyOwner {
         // TO BE IMPLEMENTED
    }
    
    function readDapi(address _dapiProxy) public view returns (uint256) {
         // TO BE IMPLEMENTED 
    }

    function validateAndPayForPaymasterTransaction  (
        bytes32,
        bytes32,
        Transaction calldata _transaction
    ) external payable onlyBootloader returns (bytes4 magic, bytes memory context) {
        // TO BE IMPLEMENTED
    }

    function postTransaction (
        bytes calldata _context,
        Transaction calldata _transaction,
        bytes32,
        bytes32,
        ExecutionResult _txResult,
        uint256 _maxRefundedGas
    ) external payable onlyBootloader override {
        // Refunds are not supported yet.
    }

    receive() external payable {}
}
```

And then proceed to go through required steps (e.g. add imports, variables, remaining logic)

Author: Nice idea, I've added the skeleton contract, and linked the OG paymasters guide too. LMK how it looks. 👍🏻

            (bool success, ) = payable(BOOTLOADER_FORMAL_ADDRESS).call{
                value: requiredETH
            }("");
            require(success, "Failed to transfer funds to the bootloader");
        } else {
            revert("Unsupported paymaster flow");
        }
    }

    function postTransaction  (
        bytes calldata _context,
        Transaction calldata _transaction,
        bytes32,
        bytes32,
        ExecutionResult _txResult,
        uint256 _maxRefundedGas
    ) onlyBootloader external payable override {
        // Refunds are not supported yet.
    }

    receive() external payable {}
}
```

## Compile and Deploy the Contracts

The script below deploys the ERC20 (mockUSDC), Greeter and the Paymaster contract. It also creates an empty wallet and mints 5k `mockUSDC` tokens for the paymaster to use at a later step. Finally, it also sends 0.05 ETH to the paymaster contract so it can pay for the transactions.

The script also calls the `setDapiProxy` to set the proxy addresses for the required dAPIs on-chain. It also sets the `greeting`.

1. Create the file `deploy-paymaster.ts` under `deploy` and copy/paste the following:

```ts

> import { utils, Wallet } from "zksync-web3";

Author: Fixed 👍🏻

import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";

> require('dotenv').config();
> 
> export default async function (hre: HardhatRuntimeEnvironment) {
>   // The wallet that will deploy the token and the paymaster
>   // It is assumed that this wallet already has sufficient funds on zkSync
>   // ⚠️ Never commit private keys to file tracking history, or your account could be compromised.
> 
>   const wallet = new Wallet(process.env.PRIVATE_KEY);

Reviewer: More of an opinionated change: 

```suggestion
require('dotenv').config();
// load wallet private key from env file
const PRIVATE_KEY = process.env.PRIVATE_KEY || "";

export default async function (hre: HardhatRuntimeEnvironment) {
  // The wallet that will deploy the token and the paymaster
  // It is assumed that this wallet already has sufficient funds on zkSync
  // ⚠️ Never commit private keys to file tracking history, or your account could be compromised.
  const wallet = new Wallet(PRIVATE_KEY);
```

Author: Makes sense, fixed 👍🏻

  const emptyWallet = Wallet.createRandom();
  console.log(`Empty wallet's address: ${emptyWallet.address}`);
  console.log(`Empty wallet's private key: ${emptyWallet.privateKey}`);

  const deployer = new Deployer(hre, wallet);

  // Deploying the ERC20 token
  const erc20Artifact = await deployer.loadArtifact("MyERC20");
  const erc20 = await deployer.deploy(erc20Artifact, ["USDC", "USDC", 18]);
  console.log(`ERC20 address: ${erc20.address}`);

  // Deploying the paymaster
  const paymasterArtifact = await deployer.loadArtifact("MyPaymaster");
  const paymaster = await deployer.deploy(paymasterArtifact, [erc20.address]);
  console.log(`Paymaster address: ${paymaster.address}`);

  // Supplying paymaster with ETH.
  await (
    await deployer.zkWallet.sendTransaction({
      to: paymaster.address,
      value: ethers.utils.parseEther("0.05"),
    })
  ).wait();

  // Setting the dAPIs in Paymaster. Head over to the API3 Market (https://market.api3.org) to verify dAPI proxy contract addresses and whether they're funded or not.
    const ETHUSDdAPI = "0x28ce555ee7a3daCdC305951974FcbA59F5BdF09b";
    const USDCUSDdAPI = "0x946E3232Cc18E812895A8e83CaE3d0caA241C2AB";

>   const setProxy = paymaster.setDapiProxy(USDCUSDdAPI, ETHUSDdAPI)

Reviewer: Nit: fix formatting

  await (await setProxy).wait()
  console.log("dAPI Proxies Set!")

  // Deploying the Greeter contract
  const greeterContractArtifact = await deployer.loadArtifact("Greeter");
  const oldGreeting = "old greeting"
  const deployGreeter = await deployer.deploy(greeterContractArtifact, [oldGreeting]);
  console.log(`Greeter contract address: ${deployGreeter.address}`);

  // Supplying the ERC20 tokens to the empty wallet:
  await // We will give the empty wallet 5k mUSDC:
  (await erc20.mint(emptyWallet.address, "5000000000000000000000")).wait();

  console.log("Minted 5k mUSDC for the empty wallet");

  console.log(`Done!`);
}
```

2. Create a `.env` file and add your private key:

```sh

> $ echo 'PRIVATE_KEY=' > .env

Reviewer: ```suggestion
echo 'PRIVATE_KEY=' > .env
```

```

3. Compile and deploy the contracts from the project root:

```sh
yarn hardhat compile
yarn hardhat deploy-zksync --script deploy-paymaster.ts
```

The output should be like this (Your values will be different):

```
Empty wallet's address: 0xcc7527d2DCb86e5327C494b323af502aEFd76831
Empty wallet's private key: 0x1d79f139605b82f3597654f274273220514ec0994fabd9f205a0a56e907d14a5
ERC20 address: 0x4CbBd2FB4700a19A19d3be5b19609f8cA6187980
Paymaster address: 0x991c592Cfc34406746b59eBA26E3D8e6f40c28bb
dAPI Proxies Set!
Greeter contract address: 0xbCC6aF86Ca5BAFedDDe922a64765Cbb438698C57
Minted 5k mUSDC for the empty wallet
Done!
```

> 4. Edit the `.env` file again to populate the following variables from the output:

Reviewer: ```suggestion
4. Update the `.env` file to populate the following variables from the above output:
```

```

> PRIVATE_KEY=

Reviewer: ```suggestion
```

PAYMASTER_ADDRESS=
TOKEN_ADDRESS=
EMPTY_WALLET_PRIVATE_KEY=
GREETER_CONTRACT=
```

:::tip
* Addresses and private keys are different on each run.
* Make sure you delete the `artifacts-zk` and `cache-zk` folders before recompiling.
:::

## Using the paymaster

1. Create the `use-paymaster.ts` script in the `deploy` folder. 

```ts
import { ContractFactory, Provider, utils, Wallet } from "zksync-web3";
import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";

> import { getDeployedContracts } from "zksync-web3/build/src/utils";

Reviewer: Unused import.
```suggestion
```

Author: Removed

require("dotenv").config();

> // Put the address of the deployed paymaster and the Greeter Contract in the .env file
> const PAYMASTER_ADDRESS = process.env.PAYMASTER_ADDRESS;
> const GREETER_CONTRACT_ADDRESS = process.env.GREETER_CONTRACT;
> 
> // Put the address of the ERC20 token in the .env file:
> const TOKEN_ADDRESS = process.env.TOKEN_ADDRESS;

Reviewer: ```suggestion
// load variables from env file
const PAYMASTER_ADDRESS = process.env.PAYMASTER_ADDRESS || "";
const GREETER_CONTRACT_ADDRESS = process.env.GREETER_CONTRACT || "";
const TOKEN_ADDRESS = process.env.TOKEN_ADDRESS || "";
const EMPTY_WALLET_PRIVATE_KEY = process.env.EMPTY_WALLET_PRIVATE_KEY || "";
```

function getToken(hre: HardhatRuntimeEnvironment, wallet: Wallet) {
  const artifact = hre.artifacts.readArtifactSync("MyERC20");
  return new ethers.Contract(TOKEN_ADDRESS, artifact.abi, wallet);
}

// Greeter contract
function getGreeter(hre: HardhatRuntimeEnvironment, wallet: Wallet) {
  const artifact = hre.artifacts.readArtifactSync("Greeter");
  return new ethers.Contract(GREETER_CONTRACT_ADDRESS, artifact.abi, wallet);
}

> // Wallet private key
> // ⚠️ Never commit private keys to file tracking history, or your account could be compromised.
> const EMPTY_WALLET_PRIVATE_KEY = process.env.EMPTY_WALLET_PRIVATE_KEY;

Reviewer: Loaded from previous suggested change.
```suggestion
```

export default async function (hre: HardhatRuntimeEnvironment) {
    const provider = new Provider("https://testnet.era.zksync.dev");

>     const emptyWallet = new Wallet(EMPTY_WALLET_PRIVATE_KEY, provider);

Reviewer: Nit: fix formatting
